### PR TITLE
Split Mask into SegmentationMask and DetectionMasks

### PR DESCRIPTION
**Pull request by pmeier**
`SegmentationMask` in singular similar to `Image`, since it can easily be batched. `DetectionMasks` in plural similar to `BoundingBoxes`, since the shape should always be `(N, H, W)` corresponding to the `(N, 4)` of `BoundingBoxes`.

All mask kernels continue to work, since we support subclasses, i.e. both `F.resize(tv_tensors.SegmentationMask(...), ...)` and `F.resize(tv_tensors.DetectionMasks(...), ...)` dispatch to `F.resize_mask`.

Some decisions I made that are up to discussion:

- Only `SegmentationMask` can be instantiated from a PIL image, since multiple detection masks cannot be stored in a single image
- `SanitizeBoundingBoxes` only sanitizes `DetectionMasks`, since there is no coupling of a potential existing `SegmentationMask` in the sample and the `BoundingBoxes` that we operate on

Documentation is TBD.

### CODE REVIEW

File: test/test_transforms_v2_refactored.py [hunk 1]
```diff
@@ -1867,7 +1872,7 @@ def test_others_catch_all_and_none(self, make_input):
         # make sure "others" works as a catch-all and that None means no conversion
 
         sample, inpt_dtype, bbox_dtype, mask_dtype = self.make_inpt_with_bbox_and_mask(make_input)
-        out = transforms.ToDtype(dtype={tv_tensors.Mask: torch.int64, "others": None})(sample)
+        out = transforms.ToDtype(dtype={tv_tensors.DetectionMasks: torch.int64, "others": None})(sample)
```
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1875]: We need to fix this in our references as well.

File: test/test_transforms_v2_refactored.py [hunk 2]
```diff
@@ -2034,8 +2040,8 @@ def test_error(self, T):
 
         for input_with_bad_type in (
             F.to_pil_image(imgs[0]),
-            tv_tensors.Mask(torch.rand(12, 12)),
-            tv_tensors.BoundingBoxes(torch.rand(2, 4), format="XYXY", canvas_size=12),
+            tv_tensors.DetectionMasks(torch.rand(12, 12)),
+            tv_tensors.BoundingBoxes(torch.rand(2, 4), format="XYXY", canvas_size=(12, 12)),
```
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:2044]: Driveby

File: torchvision/tv_tensors/_mask.py [hunk 3]
```diff
@@ -37,3 +37,34 @@ def __new__(
...
+class DetectionMasks(Mask):
+    def __new__(
+        cls,
+        data: Any,
+        *,
+        dtype: Optional[torch.dtype] = None,
+        device: Optional[Union[torch.device, str, int]] = None,
+        requires_grad: Optional[bool] = None,
+    ) -> Mask:
+        tensor = cls._to_tensor(data, dtype=dtype, device=device, requires_grad=requires_grad)
```
Comment (pmeier) [inline torchvision/tv_tensors/_mask.py:69]: We should probably have similar shape checks here as

https://github.com/pytorch/vision/blob/a06df0d9229e49bd859e2ff0355a48b3bddd1e10/torchvision/tv_tensors/_bounding_box.py#L53-L57

File: test/test_transforms_v2_refactored.py [hunk 4]
```diff
@@ -536,7 +536,8 @@ def test_functional(self, size, make_input):
             (F._resize_image_pil, PIL.Image.Image),
             (F.resize_image, tv_tensors.Image),
             (F.resize_bounding_boxes, tv_tensors.BoundingBoxes),
-            (F.resize_mask, tv_tensors.Mask),
+            (F.resize_mask, tv_tensors.SegmentationMask),
+            (F.resize_mask, tv_tensors.DetectionMasks),
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:540]: Do we need the extra parametrization? Will anything break if we ever have different kernels for SegmentationMask and DetectionMasks

File: test/test_transforms_v2_refactored.py (hunk 1 above)
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:1875]: We should not need to change this?  Passing `tv_tensors.Mask: torch.int64` should still work, right?

File: test/test_transforms_v2_refactored.py [hunk 5]
```diff
@@ -2571,7 +2582,15 @@ def test_functional_image_correctness(self, kwargs):
...
     @pytest.mark.parametrize(
         "make_input",
-        [make_image_tensor, make_image_pil, make_image, make_bounding_boxes, make_segmentation_mask, make_video],
+        [
+            make_image_tensor,
+            make_image_pil,
+            make_image,
+            make_bounding_boxes,
+            make_segmentation_mask,
+            make_detection_mask,
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:2591]: Why test both here? The other tests that were modified seem to only be `test_functional_signature`?

File: torchvision/transforms/v2/_geometry.py [hunk 6]
```diff
@@ -447,8 +447,8 @@ class Pad(Transform):
         fill (number or tuple or dict, optional): Pixel fill value used when the  ``padding_mode`` is constant.
             Default is 0. If a tuple of length 3, it is used to fill R, G, B channels respectively.
             Fill value can be also a dictionary mapping data type to the fill value, e.g.
-            ``fill={tv_tensors.Image: 127, tv_tensors.Mask: 0}`` where ``Image`` will be filled with 127 and
-            ``Mask`` will be filled with 0.
+            ``fill={tv_tensors.Image: 127, tv_tensors.SegmentationMask: 0}`` where ``Image`` will be filled with 127 and
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_geometry.py:450]: Any strong reason to remove Mask from here? Isn't this a bit limitting?

File: torchvision/transforms/v2/_misc.py [hunk 7]
```diff
@@ -415,7 +416,7 @@ def forward(self, *inputs: Any) -> Any:
 
     def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
         is_label = inpt is not None and inpt is params["labels"]
-        is_bounding_boxes_or_mask = isinstance(inpt, (tv_tensors.BoundingBoxes, tv_tensors.Mask))
+        is_bounding_boxes_or_mask = isinstance(inpt, (tv_tensors.BoundingBoxes, tv_tensors.DetectionMasks))
```
Comment (NicolasHug) [inline torchvision/transforms/v2/_misc.py:419]: We definitely don't want to do that because we don't want to force users to use `DetectionMasks`. 

Both `DetectionMasks` and `SegmentationMask` must be 100% inter-changeable with just `Mask`. Maybe that won't be the case in the future, but right now it should

File: torchvision/tv_tensors/_mask.py [hunk 8]
```diff
@@ -37,3 +37,34 @@ def __new__(
...
+
+class SegmentationMask(Mask):
+    def __new__(
+        cls,
+        data: Any,
+        *,
+        dtype: Optional[torch.dtype] = None,
+        device: Optional[Union[torch.device, str, int]] = None,
+        requires_grad: Optional[bool] = None,
+    ) -> Mask:
```
Comment (NicolasHug) [inline torchvision/tv_tensors/_mask.py:50]: annotation is wrong?

File: torchvision/tv_tensors/_mask.py [hunk 9]
```diff
@@ -37,3 +37,34 @@ def __new__(
 
         tensor = cls._to_tensor(data, dtype=dtype, device=device, requires_grad=requires_grad)
         return tensor.as_subclass(cls)
+
+
+class SegmentationMask(Mask):
+    def __new__(
```
Comment (NicolasHug) [inline torchvision/tv_tensors/_mask.py:43]: Is this `__new__` any different from the base class one?

File: torchvision/tv_tensors/_mask.py (hunk 3 above)
Comment (NicolasHug) [reply torchvision/tv_tensors/_mask.py:69]: The rationale to prevent instantiation from PIL images sounds OK to me, but we might as well raise an informative error message when passed a PIL image then

**Review by NicolasHug — COMMENTED** [review]
Thanks @pmeier , made a first pass

**Comment by pmeier** [comment]
We had an offline discussion after the initial review by @NicolasHug and there are two open points that we need to properly discuss before we move ahead here:

1. https://github.com/pytorch/vision/pull/7909#discussion_r1310442802 The dictionaries that users can supply to `dtype` for `transforms.ToDtype` and `fill` for multiple geometric transforms only support type equality matching, e.g.

   https://github.com/pytorch/vision/blob/96950a5cff2cf3ebee1b91cd3fbafe086d54c9c5/torchvision/transforms/v2/_misc.py#L264-L269

   The rationale for this behavior was to prevent `torch.Tensor` becoming the catch all key for all TVTensors. However, this also means that `dtype={tv_tensors.Mask: ...}` will not work for the proposed `tv_tensors.DetectionMasks` and `tv_tensors.SegmentationMask`.

   This is the same for user defined TVTensor subclasses. For example

    ```python
    class MyImage(tv_tensors.Image):
        ...
    
    transforms.ToDtype(dtype={tv_tensors.Image: ...})  # does not work for MyImage
    ```

    There were two suggestions on how to solve this:

    1. We could implement a MRO traversal similar to what we do for the kernel selection.
    2. We could just alias `SegmentationMask` and `DetectionMasks` to `Mask`.

    We didn't go into i., but for ii. there was some concern regarding the usefulness. The idea for all of this was to make it less confusing for users that `Mask` represents two related but still different concepts. But the story "you can use DetectionMasks and SegmentationMask but internally we treat them exactly as Mask" does not sound less confusing than "just use Mask for detection and segmentation tasks".
2. https://github.com/pytorch/vision/pull/7909#discussion_r1310450898. There was some disagreement if `SegmentationMasks` and `DetectionMask` need to both function as `Mask` does now. Right now `Mask` works in segmentation and detection contexts, since we have no way to distinguish them based on the mask object. 
   But if we do, as proposed in this PR, we can capitalize on that. For example `SanitizeBoundingBoxes` should not touch `SegmentationMask`, but only `DetectionMasks`, since the former has no connection to the `BoundingBoxes`.

Since the branch for the upcoming release is already cut (#7917) and the points above need thorough discussion, I'm going to close this for now. Since we'll promote the v2 transforms as stable-ish, we'll have fewer options in all of this in the future. Still, this is better than rushing this today and end up with something worse.
